### DISCUSSION TITLE
Issues running the automation - function IPRAP and K3S Nodes Job

### DISCUSSION THREAD

**Discussion by RodCSouza** [techno-tim/k3s-ansible#142] [Questions]
Hello team,

I'm trying to use the automation, but I've faced two issues:

Looks like that the adjustments made to cover IPV6 causes a issue in the determination of the IP for the parameter "apiserver_endpoint".

My scenario is one environment where there is no DHCP for IPV6, so the implemented use of function "ansible.utils.ipwrap", returns no data and consequently no IP address information, to correct the errors and be able to continue the automation execution, I've commented the lines with this functions in the file ks3s>master>tasks>main.yml (line 163), this solved the step "Configure kubectl cluster to".

But now I'm facing an error in the step node "Enable and check K3s service", the resultant file of the copy from the template "k3s.service.j2", has also the use of the function "ansible.utils.ipwrap", so I've commented that as well, but the main issue on this point is that the file in the node has no values substituting the corresponding variables, resulting into a job error when trying to rump up the service.

The file content is like this:
`
-- The job identifier is 1862 and the job result is failed.
alchemist@k3s2:/etc/systemd/system$ cat k3s-node.service
[Unit]
Description=Lightweight Kubernetes
Documentation=https://k3s.io
After=network-online.target

[Service]
Type=notify
ExecStartPre=-/sbin/modprobe br_netfilter
ExecStartPre=-/sbin/modprobe overlay
ExecStart=/usr/local/bin/k3s agent --server https://{{ apiserver_endpoint }}:6443 --token {{ hostvars[groups['master'][0]]['token'] | default(k3s_token) }} {{ extra_agent_args | default("") }}
`
I'm wondering if there is a need of Defaults folder and main.yml, for the Nodes section, I'm far from a Ansible expert, so it is just a guess.

Would be nice if you could help me with this.

Thanks.

**Comment by sleiner** [comment]
> My scenario is one environment where there is no DHCP for IPV6, so the implemented use of function "ansible.utils.ipwrap", returns no data and consequently no IP address information

It seems to me like you’re conflating a few things here: The ipwrap filter‘s only purpose is [wrapping IPv6 addresses (and only IPv6 addresses) in square brackets](https://docs.ansible.com/ansible/latest/collections/ansible/utils/docsite/filters_ipaddr.html#wrapping-ipv6-addresses-in-brackets). So the output of ipwrap does  only depend on the input string, not on the network situation (e.g. the existence of a DHCP6 server).

What is the error message that you are gering with the _unmodified_ roles?

**Reply by RodCSouza** [reply]
Hello Sleiner,

Without the comment on the Main.yml of the Master group, this is the error message generated in the step Configure kubectl cluster to:
`
fatal: [10.75.15.105]: FAILED! => {"changed": true, "cmd": ["k3s", "kubectl", "config", "set-cluster", "default", "--server=https://{{", "apiserver_endpoint", "|", "ansible.utils.ipwrap", "}}:6443", "--kubeconfig", "~alchemist/.kube/config"], "delta": "0:00:00.180137", "end": "2022-11-04 04:00:24.374003", "msg": "non-zero return code", "rc": 1, "start": "2022-11-04 04:00:24.193866", "stderr": "error: Unexpected args: [default apiserver_endpoint | ansible.utils.ipwrap }}:6443]", "stderr_lines": ["error: Unexpected args: [default apiserver_endpoint | ansible.utils.ipwrap }}:6443]"], "stdout": "Set a cluster entry in kubeconfig.\n\n Specifying a name that already exists will merge new fields on top of existing values for those fields.\n\nExamples:\n  # Set only the server field on the e2e cluster entry without touching other values\n  kubectl config set-cluster e2e --server=https://1.2.3.4\n  \n  # Embed certificate authority data for the e2e cluster entry\n  kubectl config set-cluster e2e --embed-certs --certificate-authority=~/.kube/e2e/kubernetes.ca.crt\n  \n  # Disable cert checking for the e2e cluster entry\n  kubectl config set-cluster e2e --insecure-skip-tls-verify=true\n  \n  # Set custom TLS server name to use for validation for the e2e cluster entry\n  kubectl config set-cluster e2e --tls-server-name=my-cluster-name\n  \n  # Set proxy url for the e2e cluster entry\n  kubectl config set-cluster e2e --proxy-url=https://1.2.3.4\n\nOptions:\n    --embed-certs=false:\n\tembed-certs for the cluster entry in kubeconfig\n\n    --proxy-url='':\n\tproxy-url for the cluster entry in kubeconfig\n\nUsage:\n  kubectl config set-cluster NAME [--server=server] [--certificate-authority=path/to/certificate/authority] [--insecure-skip-tls-verify=true] [--tls-server-name=example.com] [options]\n\nUse \"kubectl options\" for a list of global command-line options (applies to all commands).", "stdout_lines": ["Set a cluster entry in kubeconfig.", "", " Specifying a name that already exists will merge new fields on top of existing values for those fields.", "", "Examples:", "  # Set only the server field on the e2e cluster entry without touching other values", "  kubectl config set-cluster e2e --server=https://1.2.3.4", "  ", "  # Embed certificate authority data for the e2e cluster entry", "  kubectl config set-cluster e2e --embed-certs --certificate-authority=~/.kube/e2e/kubernetes.ca.crt", "  ", "  # Disable cert checking for the e2e cluster entry", "  kubectl config set-cluster e2e --insecure-skip-tls-verify=true", "  ", "  # Set custom TLS server name to use for validation for the e2e cluster entry", "  kubectl config set-cluster e2e --tls-server-name=my-cluster-name", "  ", "  # Set proxy url for the e2e cluster entry", "  kubectl config set-cluster e2e --proxy-url=https://1.2.3.4", "", "Options:", "    --embed-certs=false:", "\tembed-certs for the cluster entry in kubeconfig", "", "    --proxy-url='':", "\tproxy-url for the cluster entry in kubeconfig", "", "Usage:", "  kubectl config set-cluster NAME [--server=server] [--certificate-authority=path/to/certificate/authority] [--insecure-skip-tls-verify=true] [--tls-server-name=example.com] [options]", "", "Use \"kubectl options\" for a list of global command-line options (applies to all commands)."]}
fatal: [10.75.15.104]: FAILED! => {"changed": true, "cmd": ["k3s", "kubectl", "config", "set-cluster", "default", "--server=https://{{", "apiserver_endpoint", "|", "ansible.utils.ipwrap", "}}:6443", "--kubeconfig", "~alchemist/.kube/config"], "delta": "0:00:00.183015", "end": "2022-11-04 04:00:24.386876", "msg": "non-zero return code", "rc": 1, "start": "2022-11-04 04:00:24.203861", "stderr": "error: Unexpected args: [default apiserver_endpoint | ansible.utils.ipwrap }}:6443]", "stderr_lines": ["error: Unexpected args: [default apiserver_endpoint | ansible.utils.ipwrap }}:6443]"], "stdout": "Set a cluster entry in kubeconfig.\n\n Specifying a name that already exists will merge new fields on top of existing values for those fields.\n\nExamples:\n  # Set only the server field on the e2e cluster entry without touching other values\n  kubectl config set-cluster e2e --server=https://1.2.3.4\n  \n  # Embed certificate authority data for the e2e cluster entry\n  kubectl config set-cluster e2e --embed-certs --certificate-authority=~/.kube/e2e/kubernetes.ca.crt\n  \n  # Disable cert checking for the e2e cluster entry\n  kubectl config set-cluster e2e --insecure-skip-tls-verify=true\n  \n  # Set custom TLS server name to use for validation for the e2e cluster entry\n  kubectl config set-cluster e2e --tls-server-name=my-cluster-name\n  \n  # Set proxy url for the e2e cluster entry\n  kubectl config set-cluster e2e --proxy-url=https://1.2.3.4\n\nOptions:\n    --embed-certs=false:\n\tembed-certs for the cluster entry in kubeconfig\n\n    --proxy-url='':\n\tproxy-url for the cluster entry in kubeconfig\n\nUsage:\n  kubectl config set-cluster NAME [--server=server] [--certificate-authority=path/to/certificate/authority] [--insecure-skip-tls-verify=true] [--tls-server-name=example.com] [options]\n\nUse \"kubectl options\" for a list of global command-line options (applies to all commands).", "stdout_lines": ["Set a cluster entry in kubeconfig.", "", " Specifying a name that already exists will merge new fields on top of existing values for those fields.", "", "Examples:", "  # Set only the server field on the e2e cluster entry without touching other values", "  kubectl config set-cluster e2e --server=https://1.2.3.4", "  ", "  # Embed certificate authority data for the e2e cluster entry", "  kubectl config set-cluster e2e --embed-certs --certificate-authority=~/.kube/e2e/kubernetes.ca.crt", "  ", "  # Disable cert checking for the e2e cluster entry", "  kubectl config set-cluster e2e --insecure-skip-tls-verify=true", "  ", "  # Set custom TLS server name to use for validation for the e2e cluster entry", "  kubectl config set-cluster e2e --tls-server-name=my-cluster-name", "  ", "  # Set proxy url for the e2e cluster entry", "  kubectl config set-cluster e2e --proxy-url=https://1.2.3.4", "", "Options:", "    --embed-certs=false:", "\tembed-certs for the cluster entry in kubeconfig", "", "    --proxy-url='':", "\tproxy-url for the cluster entry in kubeconfig", "", "Usage:", "  kubectl config set-cluster NAME [--server=server] [--certificate-authority=path/to/certificate/authority] [--insecure-skip-tls-verify=true] [--tls-server-name=example.com] [options]", "", "Use \"kubectl options\" for a list of global command-line options (applies to all commands)."]}
`

The Ansible screen log also displays the task title without variables substitutions:
"TASK [k3s/master : Configure kubectl cluster to https://{{ apiserver_endpoint | ansible.utils.ipwrap }}:6443]"

With my change the log shows the tittle with the substitutions and there is no error in the task as well:
"TASK [k3s/master : Configure kubectl cluster to https://10.75.15.5:6443]"

In the step that I'm facing the error with the removal of the function "ansible.utils.ipwrap" is in the task "Copy K3s service file" for the Node group, resulting that in the node copied file, the variables has not been replaced by the values, like the "apiserver_endpoint" and "hostvars[groups['master'][0]]['token'".

**Reply by sleiner** [reply]
Interesting... What version of Ansible are you running?

**Reply by RodCSouza** [reply]
From the executing machine:
![image](https://user-images.githubusercontent.com/96396078/199994468-79e1dcce-0b21-4b4c-8704-a004ed46ca43.png)

**Reply by sleiner** [reply]
Okay... That version of Ansible should support the `ipwrap` filter (although it is not exactly new 😅). Just to make sure that the filter is not somehow broken on your system: What happens when you run `ansible localhost -m debug -a "msg='{{ [\"fd::\", \"12.34.56.78\"] | ansible.utils.ipwrap }}'"`?

**Reply by RodCSouza** [reply]
[WARNING]: Unable to parse /home/xxxxxxxx/k3s-ansible/hosts.ini as an inventory source
[WARNING]: No inventory was parsed, only implicit localhost is available
localhost | SUCCESS => {
    "msg": "{{ [\"fd::\", \"12.34.56.78\"] | ansible.utils.ipwrap }}"
}

Looks like the module is not installed, I'm checking.

**Reply by RodCSouza** [reply]
I would say that the root cause of the issue is related to one of the very first steps of the implementation:
`ansible-galaxy install -r ./collections/requirements.yml`

This one did nothing in the system, not even if run with sudo, and none of the itens in the collection list had been installed, running the command the system just accepted, without errors or any other information in the command line.

I've installed each of then manually, following the template below, changing the last section with each item of the collection.
`ansible-galaxy collection install ansible.utils`

**Reply by RodCSouza** [reply]
Was exactly this, root cause was the installation of "requirements", after installing each package manually, everything run smoothly and I could get K3S HA running successfully.

I would suggest one improvement in the playbook, to include in the tasks, steps to check the availability for each item on the requirements list.
